spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java

```diff
@@ -378,7 +378,19 @@ public SftpSession getSession() {
 				jschSession = new JSchSessionWrapper(initJschSession());
 			}
 			SftpSession sftpSession = new SftpSession(jschSession);
-			sftpSession.connect();
+
+			if (this.isSharedSession) {
+				this.sharedSessionLock.readLock().lock();
+			}
+			try {
+				sftpSession.connect();
```

Comment:
Perhaps it would be better to just connect the jschSession within the lock above (and add a catch); then the connect within SftpSession would be a no-op for shared sessions and we'd avoid this extra lock() and the try/finally for cases when a shared session is not being used.